plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftConnector.java

```diff
@@ -0,0 +1,760 @@
+    private static final Long UID_USER = Long.valueOf(1000);
+
+    private final OpenShiftClient    openShiftClient;
+    private final KubernetesLabelConverter kubernetesLabelConverter;
+    private final KubernetesEnvVar kubernetesEnvVar;
+    private final KubernetesContainer kubernetesContainer;
+    private final KubernetesService kubernetesService;
+    private final String             cheOpenShiftProjectName;
+    private final String             cheOpenShiftServiceAccount;
+
+    @Inject
+    public OpenShiftConnector(DockerConnectorConfiguration connectorConfiguration,
+                              DockerConnectionFactory connectionFactory,
+                              DockerRegistryAuthResolver authResolver,
+                              DockerApiVersionPathPrefixProvider dockerApiVersionPathPrefixProvider,
+                              @Named("che.openshift.endpoint") String openShiftApiEndpoint,
+                              @Named("che.openshift.username") String openShiftUserName,
+                              @Named("che.openshift.password") String openShiftUserPassword,
+                              @Named("che.openshift.project") String cheOpenShiftProjectName,
+                              @Named("che.openshift.serviceaccountname") String cheOpenShiftServiceAccount) {
+        super(connectorConfiguration, connectionFactory, authResolver, dockerApiVersionPathPrefixProvider);
+        this.cheOpenShiftProjectName = cheOpenShiftProjectName;
+        this.cheOpenShiftServiceAccount = cheOpenShiftServiceAccount;
+
+        Config config = new ConfigBuilder().withMasterUrl(openShiftApiEndpoint)
+                .withUsername(openShiftUserName)
+                .withPassword(openShiftUserPassword).build();
+        this.openShiftClient = new DefaultOpenShiftClient(config);
+
+        this.kubernetesLabelConverter = new KubernetesLabelConverter();
+        this.kubernetesEnvVar = new KubernetesEnvVar();
+        this.kubernetesContainer = new KubernetesContainer();
+        this.kubernetesService = new KubernetesService();
+    }
+
+    /**
+     * @param createContainerParams
+     * @return
+     * @throws IOException
+     */
+    @Override
+    public ContainerCreated createContainer(CreateContainerParams createContainerParams) throws IOException {
+        String containerName = getNormalizedContainerName(createContainerParams);
+        String workspaceID = getCheWorkspaceId(createContainerParams);
+        // Generate workspaceID if CHE_WORKSPACE_ID env var does not exist
+        workspaceID = workspaceID.isEmpty() ? generateWorkspaceID() : workspaceID;
+        String imageName = createContainerParams.getContainerConfig().getImage();
+        Set<String> containerExposedPorts = createContainerParams.getContainerConfig().getExposedPorts().keySet();
+        Set<String> imageExposedPorts = inspectImage(imageName).getConfig().getExposedPorts().keySet();
+        Set<String> exposedPorts = new HashSet<>();
```

Comment:
hmmm.... I have extracted the logic to `getExposedPorts` method. However, I did not use constructor for setting initial capacity. I do believe using default capacity=16 and locaFactor=0.75 should work just fine for our case. If the capacity will be set it will result in increasing buckets sizes x2 after 'addAll' call